// scandit-sdk/src/lib/cameraManager.ts
import { Camera } from "./camera";
import { CameraSettings } from "./cameraSettings";

/**
 * @hidden
 */
export abstract class CameraManager {
  public selectedCamera?: Camera;
  public activeCamera?: Camera;
  public activeCameraSettings?: CameraSettings;

  public abstract setInteractionOptions(
    cameraSwitcherEnabled: boolean,
    torchToggleEnabled: boolean,
    tapToFocusEnabled: boolean,
    pinchToZoomEnabled: boolean
  ): void;

  public abstract isCameraSwitcherEnabled(): boolean;

  public abstract setCameraSwitcherEnabled(enabled: boolean): Promise<void>;

  public abstract isTorchToggleEnabled(): boolean;

  public abstract setTorchToggleEnabled(enabled: boolean): void;

  public abstract isTapToFocusEnabled(): boolean;

  public abstract setTapToFocusEnabled(enabled: boolean): void;

  public abstract isPinchToZoomEnabled(): boolean;

  public abstract setPinchToZoomEnabled(enabled: boolean): void;

  public abstract setSelectedCamera(camera?: Camera): void;

  public abstract setSelectedCameraSettings(cameraSettings?: CameraSettings): void;

  public abstract setupCameras(): Promise<void>;

  public abstract stopStream(): void;

  public abstract applyCameraSettings(cameraSettings?: CameraSettings): Promise<void>;

  public abstract reinitializeCamera(): void;

  public abstract initializeCameraWithSettings(camera: Camera, cameraSettings?: CameraSettings): Promise<void>;

  public abstract setTorchEnabled(enabled: boolean): void;

  public abstract toggleTorch(): void;

  public abstract setZoom(zoomPercentage: number, currentZoom?: number): void;
}
